File: evita/dom/forms/radio_button_control.cc
// Copyright (c) 2014 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "evita/dom/forms/radio_button_control.h"

#include "evita/dom/events/form_event.h"
#include "evita/dom/forms/form.h"
#include "evita/dom/forms/form_control.h"

namespace dom {

//////////////////////////////////////////////////////////////////////
//
// RadioButtonControl
//
RadioButtonControl::RadioButtonControl(const base::string16& name)
    : ScriptableBase(name), checked_(false) {}

RadioButtonControl::~RadioButtonControl() {}

void RadioButtonControl::set_checked(bool new_checked) {
  if (checked_ == new_checked)
    return;
  checked_ = new_checked;
}

// FormControl
std::unique_ptr<domapi::FormControl> RadioButtonControl::Paint(
    const FormPaintInfo& paint_info) const {
  auto state = ComputeState(paint_info);
  if (checked_)
    state.set_checked();
  return std::make_unique<domapi::RadioButton>(event_target_id(), bounds(),
                                               state);
}

}  // namespace dom

File: evita/dom/forms/radio_button_control.js
// Copyright (c) 2014 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

(function() {
  /**
   * @this {!RadioButtonControl}
   * @return {boolean}
   */
  function getChecked() {
    return this.checked_;
  }

  /**
   * @this {!RadioButtonControl}
   * @param {boolean} newChecked
   */
  function setChecked(newChecked) {
    if (this.checked_ === newChecked)
      return;
    this.dispatchChangeEvent();
    this.checked_ = newChecked;
    if (!this.form || !newChecked)
      return;
    // Turn checked off for other radio button in same radio button group.
    for (const other of this.form.controls) {
      if (other === this)
        continue;
      if (!(other instanceof RadioButtonControl))
        continue;
      if (other.name !== this.name)
        continue;
      if (!other.checked)
        continue;
      other.checked_ = false;
      other.dispatchChangeEvent();
      return;
    }
  }

  Object.defineProperties(RadioButtonControl.prototype, {
    checked: {get: getChecked, set: setChecked},
  });

  /**
   * @this {!RadioButtonControl}
   * @param {!Event} event
   * Default event handler.
   */
  function handleEvent(event) {
    var radioButton = /** @type {!RadioButtonControl} */(event.target);
    switch (event.type) {
      case Event.Names.CLICK:
        this.dispatchInputEvent();
        radioButton.checked = !radioButton.checked;
        radioButton.focus();
        break;
    }
    FormControl.handleEvent.call(this, event);
  }

  Object.defineProperties(RadioButtonControl, {
    handleEvent: {value: handleEvent},
  });
})();
